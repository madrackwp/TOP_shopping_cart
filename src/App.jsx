import { useState, useEffect, useLayoutEffect } from "react";
import { createBrowserRouter, RouterProvider } from "react-router-dom";
import Home from "./Home";
import Shop from "./Shop";
import "./styles/index.css";
import Layout from "./Layout";

function App() {
  const [cartItems, setCartItems] = useState({});

  const [shopData, setShopData] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  function checkout() {
    let checkoutMessage = "You are checking out with:\n";
    const cartEntries = Object.entries(cartItems);
    cartEntries.forEach((value, key) => {
      // console.log(value, key);
      console.log(value[1].count);
      console.log(value[1].item.title);
      checkoutMessage =
        checkoutMessage + value[1].count + " x " + value[1].item.title + "\n";
    });

    alert(checkoutMessage);
  }

  function removeItemFromCart(itemID) {
    console.log(`Trying to remove: ${itemID}`);
    const oldCartItems = { ...cartItems };

    delete oldCartItems[itemID];

    setCartItems(oldCartItems);
  }

  function updateCart(updateData) {
    const oldCartItems = { ...cartItems };
    //We check if updateData exists in the oldCartItems
    if (updateData.id in oldCartItems) {
      oldCartItems[updateData.id].count += 1;
    } else {
      oldCartItems[updateData.id] = { item: updateData, count: 1 };
    }
    // oldCartItems.push(updateData);
    setCartItems(oldCartItems);
  }

  useEffect(() => {
    // Function to fetch data from the API
    const fetchData = async () => {
      try {
        const response = await fetch(
          "https://fakestoreapi.com/products?limit=15"
        );
        if (!response.ok) {
          throw new Error("Network response was not ok");
        }
        const result = await response.json();
        setShopData(result);
        // console.log("Shop data received!");
        // console.log(result);
      } catch (error) {
        setError(error.message);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
    // console.log(shopData);
  }, []);

  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return <div>Error: {error}</div>;
  }

  const router = createBrowserRouter([
    {
      path: "/",
      element: <Layout itemsInCart={cartItems} />,
      children: [
        {
          path: "/",
          element: <Home />,
        },
        {
          path: "/shop",
          element: (
            <Shop
              shopData={shopData}
              cartData={cartItems}
              updateCartHandler={updateCart}
              cartDeleteHandler={removeItemFromCart}
              checkoutHandler={checkout}
            />
          ),
        },
      ],
    },
  ]);

  return <RouterProvider router={router} />;
}

export default App;
