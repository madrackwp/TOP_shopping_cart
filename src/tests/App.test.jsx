import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import App from "../App";

describe("App", () => {
  it("renders headline", () => {
    render(<App title="React" />);

    // screen.debug();

    // check if App components renders headline
  });

  it("renders magnificent monkeys", () => {
    // since screen does not have the container property, we'll destructure render to obtain a container for this test
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
});

// describe("The shop name should be at the top", () => {
//   it("renders correct heading", () => {
//     render(<App />);
//     expect(screen.getByRole("heading").textContent).toMatch(/racky's wares/i);
//   });
// });

it("renders radical rhinos after button click", async () => {
  const user = userEvent.setup();

  render(<App />);
  const button = screen.getByRole("button", { name: "Shop" });

  await user.click(button);

  expect(screen.getByRole("heading").textContent).toMatch(
    /this is the shop page/i
  );
});
