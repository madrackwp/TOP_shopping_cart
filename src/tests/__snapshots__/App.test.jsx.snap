// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders magnificent monkeys 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;
